rust-symcrypt/src/rsa.rs

```diff
@@ -0,0 +1,578 @@
+/// needed to call SymCrypt since the same data is stored on the state or is assumed
+/// 2. to keep this thread safe, I've combined the alloc / generate and alloc / set pattern to just one function
+/// 3. This code is pretty raw, and would not be very usable, the idea would be to write a crate on top of this one, called something like symcrypt_rsa_parse, which will be a friendlier
+/// interface on top of the RSA code we have right now it would have convenience functions like rsa_from_pkcs8, rsa_from_der, rsa_from_asn1 etc. ( this could be part of the intern project as well.)
+/// 4. with how many steps / options there is for generating a key, we can maybe do a builder pattern, this would be similar to how it is on NCrypt / BCrypt where they have
+/// RsaKey functions and the RsaKey.finalize() at the end.
+/// ex:
+///         builder = RsaKeyBuilder::new()
+///             .key_mode(RsaKeyMode::KeyPair)
+///             .key_usage(RsaKeyUsage::Sign)
+///             .generate_key()
+/// My thought is to leave it how it is right now as there are not that many parameters that we need to specify, but thought I would point out the pattern.
+///
+/// 5.
+// !REVIEW: instead of using in/out buffers for the get functions, we can use Vec<> which will be slower but more robust, and will not require the user to know the len of the fields,
+//  Or we could instead return a struct with the info
+//       pub struct KeyReturnValue {
+//          key_mode : RsaKeyMode,
+//          modulus: u64,
+//          p: &[u8],
+//          q: &[u8],
+//              ...
+//        }
+
+/// 6. A lot of the functions take in a NumberFormat ( MSB, LSB.) Should we include this when we create a new Public / Key pair state?  Also talked with Phil, ASN is MSB, are there any
+/// use cases for LSB? Should we remove the option from the user and force MSB?
+///
+// Rsa functions. For further documentation please refer to symcrypt.h
+// TODO rest of the documentation
+use crate::errors::SymCryptError;
+use crate::NumberFormat;
+use std::ptr;
+
+/// 2^16 + 1
+const DEFAULT_PUBLIC_EXPONENT: u64 = 0x10001;
+
+/// Rsa Public and Private Key State.
+///
+/// [`RsaKeyPair`] stores a private and public Rsa key.
+pub struct RsaKeyPair {
+    inner: symcrypt_sys::PSYMCRYPT_RSAKEY,
+    key_usage: RsaKeyUsage,
+    number_format: NumberFormat,
+}
+
+/// Rsa Public Key State
+///
+/// [`RsaPublicKey`] stores an Rsa public key.
+pub struct RsaPublicKey {
+    inner: symcrypt_sys::PSYMCRYPT_RSAKEY,
+    key_usage: RsaKeyUsage,
+    number_format: NumberFormat,
+}
+
+#[derive(Debug, Copy, Clone, PartialEq)]
+/// `RsaKeyUsage` will indicate if the [`RsaKeyPair`] or [`RsaPublicKey`] will be used for [`RsaKeyUsage::Sign`] or [`RsaKeyUsage::Encrypt`].
+pub enum RsaKeyUsage {
+    Sign,
+    Encrypt,
+}
```

Comment:
In symcrypt both can be set at the same time. The caller may need to do both. Enrolling for an encryption cert. Also need to do a one time sign.

Need to add SignAndEncrypt choice.

Reply:
Another option to follow bcryot, Encrypt should set both bits. Sign only set the Sign bit.

Reply:
From a conversation that i had with Sam, I thought that the semantic of Sign + Verify is a grey area for FIPS.

Reply:
Not a grey area. BCrypt, also sets Sign for Encrypt. You need to look at the BCrypt code where it calls these Symcrypt APIs